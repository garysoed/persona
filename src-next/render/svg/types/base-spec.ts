import {ObservableOrValue} from '../../types/observable-or-value';

export interface BaseSpec {
  readonly children: ObservableOrValue<readonly BaseSpec[]>;
}
