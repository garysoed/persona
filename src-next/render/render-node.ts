import {Observable} from 'rxjs';

import {renderBase} from './render-base';
import {NodeWithId} from './types/node-with-id';
import {RenderNodeSpec} from './types/render-node-spec';


export function renderNode<N extends Node>(spec: RenderNodeSpec<N>): Observable<NodeWithId<N>> {
  return renderBase(spec, spec.node);
}
