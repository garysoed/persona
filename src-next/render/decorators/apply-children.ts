import {$asArray, $filterNonNull, $pipe, arrayFrom} from 'gs-tools/export/collect';
import {diffArray} from 'gs-tools/export/rxjs';
import {combineLatest, Observable, of as observableOf} from 'rxjs';
import {map, switchMap, tap} from 'rxjs/operators';

import {render} from '../render';
import {NodeWithId} from '../types/node-with-id';
import {RenderContext} from '../types/render-context';
import {RenderSpec} from '../types/render-spec';

import {Decorator} from './apply-decorators';


export function applyChildren(
    children$: Observable<readonly RenderSpec[]>,
    context: RenderContext,
): Decorator<NodeWithId<Element|DocumentFragment>> {
  return el => children$.pipe(
      switchMap(specs => {
        const renderedNode$list = specs.map(spec => render(spec, context));
        if (renderedNode$list.length <= 0) {
          return observableOf([]);
        }

        return combineLatest(renderedNode$list).pipe(
            map(renderedNodes => $pipe(renderedNodes, $filterNonNull(), $asArray())),
        );
      }),
      diffArray(),
      tap(diff => {
        switch (diff.type) {
          case 'delete':
            el.removeChild(diff.value);
            break;
          case 'init':
            for (const child of arrayFrom(el.children)) {
              el.removeChild(child);
            }

            for (const child of diff.value) {
              el.appendChild(child);
            }
            break;
          case 'insert':{
            const insertBefore = el.children.item(diff.index);
            el.insertBefore(diff.value, insertBefore);
            break;
          }
          case 'set':{
            const toDelete = el.children.item(diff.index);
            const setBefore = toDelete?.nextSibling || null;
            if (toDelete) {
              el.removeChild(toDelete);
            }
            el.insertBefore(diff.value, setBefore);
            break;
          }
        }
      }),
  );
}