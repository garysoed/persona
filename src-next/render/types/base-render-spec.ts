import {Decorator} from '../decorators/apply-decorators';

import {NodeWithId} from './node-with-id';

export interface BaseRenderSpec<N extends Node> {
  readonly id: unknown;
  readonly decorators?: ReadonlyArray<Decorator<NodeWithId<N>>>;
}