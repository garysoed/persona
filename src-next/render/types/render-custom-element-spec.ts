import {getOwnPropertyKeys} from 'gs-tools/export/typescript';
import {Observable, of as observableOf} from 'rxjs';

import {Binding, Spec, UnresolvedBindingSpec} from '../../types/ctrl';
import {Registration} from '../../types/registration';

import {BaseRenderSpec} from './base-render-spec';
import {normalize, normalizeMap, ObservableOrValue} from './observable-or-value';
import {RenderSpec} from './render-spec';
import {RenderSpecType} from './render-spec-type';


/**
 * Values of the input for the given spec.
 *
 * @remarks
 * The values must be `Observable`. They must emit whenever the value changes to update the values
 * of the custom element.
 *
 * @thHidden
 */
export type InputsOf<S extends UnresolvedBindingSpec> = Partial<{
  readonly [K in keyof S]: Binding<S[K]> extends Observable<infer T> ? ObservableOrValue<T> : never;
}>;

export type NormalizedInputsOf<S extends UnresolvedBindingSpec> = Partial<{
  readonly [K in keyof S]: Binding<S[K]> extends Observable<infer T> ? Observable<T> : never;
}>;

interface InputRenderSpec<S extends Spec> extends BaseRenderSpec<HTMLElement> {
  readonly registration: Registration<HTMLElement, S>;
  readonly attrs?: ReadonlyMap<string, ObservableOrValue<string|undefined>>;
  readonly children?: ObservableOrValue<readonly RenderSpec[]>;
  readonly inputs: InputsOf<S['host']&{}>;
  readonly styles?: ObservableOrValue<ReadonlyMap<string, string|null>>;
  readonly textContent?: ObservableOrValue<string>;
}

export interface RenderCustomElementSpec<S extends Spec> extends InputRenderSpec<S> {
  readonly type: RenderSpecType.CUSTOM_ELEMENT;
  readonly attrs?: ReadonlyMap<string, Observable<string|undefined>>;
  readonly children?: Observable<readonly RenderSpec[]>;
  readonly inputs: NormalizedInputsOf<S['host']&{}>;
  readonly styles?: Observable<ReadonlyMap<string, string|null>>;
  readonly textContent?: Observable<string>;
}

export function renderCustomElement<S extends Spec>(input: InputRenderSpec<S>): RenderCustomElementSpec<S> {
  return {
    ...input,
    type: RenderSpecType.CUSTOM_ELEMENT,
    attrs: input.attrs ? normalizeMap(input.attrs) : undefined,
    children: input.children ? normalize(input.children) : undefined,
    inputs: normalizedInputs(input.inputs),
    styles: input.styles ? normalize(input.styles) : undefined,
    textContent: input.textContent !== undefined ? normalize(input.textContent) : undefined,
  };
}

function normalizedInputs<S extends UnresolvedBindingSpec>(spec: InputsOf<S>): NormalizedInputsOf<S> {
  const output: {[K in keyof S]?: Observable<unknown>} = {};
  for (const key of getOwnPropertyKeys(spec)) {
    const v: ObservableOrValue<unknown> = spec[key];
    output[key] = v instanceof Observable ? v : observableOf(v);
  }
  return output as NormalizedInputsOf<S>;
}