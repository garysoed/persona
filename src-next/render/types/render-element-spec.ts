import {Observable} from 'rxjs';

import {BaseRenderSpec} from './base-render-spec';
import {normalize, normalizeMap, ObservableOrValue} from './observable-or-value';
import {RenderSpec} from './render-spec';
import {RenderSpecType} from './render-spec-type';

interface Input extends BaseRenderSpec<HTMLElement> {
  readonly tag: string;
  readonly attrs?: ReadonlyMap<string, ObservableOrValue<string|undefined>>;
  readonly children?: ObservableOrValue<readonly RenderSpec[]>;
  readonly styles?: ObservableOrValue<ReadonlyMap<string, string|null>>;
  readonly textContent?: ObservableOrValue<string|undefined>;
}

export interface RenderElementSpec extends Input {
  readonly type: RenderSpecType.ELEMENT;
  readonly attrs?: ReadonlyMap<string, Observable<string|undefined>>;
  readonly children?: Observable<readonly RenderSpec[]>;
  readonly styles?: Observable<ReadonlyMap<string, string|null>>;
  readonly textContent?: Observable<string|undefined>;
}

export function renderElement(input: Input): RenderElementSpec {
  return {
    ...input,
    type: RenderSpecType.ELEMENT,
    attrs: input.attrs ? normalizeMap(input.attrs) : undefined,
    children: input.children ? normalize(input.children) : undefined,
    styles: input.styles ? normalize(input.styles) : undefined,
    textContent: input.textContent !== undefined ? normalize(input.textContent) : undefined,
  };
}