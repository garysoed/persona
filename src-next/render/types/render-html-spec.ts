import {Observable} from 'rxjs';

import {ParseType} from '../html-parse-service';

import {BaseRenderSpec} from './base-render-spec';
import {normalize, ObservableOrValue} from './observable-or-value';
import {RenderSpecType} from './render-spec-type';


interface Input extends BaseRenderSpec<Element> {
  readonly raw: ObservableOrValue<string>;
  readonly parseType: ParseType;
}

export interface RenderHtmlSpec extends Input {
  readonly type: RenderSpecType.HTML;
  readonly raw: Observable<string>
}

export function renderHtml(input: Input): RenderHtmlSpec {
  return {
    ...input,
    raw: normalize(input.raw),
    type: RenderSpecType.HTML,
  };
}
