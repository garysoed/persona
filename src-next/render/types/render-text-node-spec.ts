import {Observable} from 'rxjs';

import {BaseRenderSpec} from './base-render-spec';
import {normalize, ObservableOrValue} from './observable-or-value';
import {RenderSpecType} from './render-spec-type';


interface Input extends BaseRenderSpec<Text> {
  readonly textContent: ObservableOrValue<string>;
}

export interface RenderTextNodeSpec extends Input {
  readonly type: RenderSpecType.TEXT_NODE;
  readonly textContent: Observable<string>;
}

export function renderTextNode(input: Input): RenderTextNodeSpec {
  return {
    ...input,
    type: RenderSpecType.TEXT_NODE,
    textContent: normalize(input.textContent),
  };
}
