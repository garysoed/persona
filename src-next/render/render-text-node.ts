import {defer, Observable} from 'rxjs';

import {Decorator} from './decorators/apply-decorators';
import {applyTextContent} from './decorators/apply-text-content';
import {renderNode} from './render-node';
import {NodeWithId} from './types/node-with-id';
import {RenderSpecType} from './types/render-spec-type';
import {RenderTextNodeSpec} from './types/render-text-node-spec';


export function renderTextNode(
    spec: RenderTextNodeSpec,
    document: Document,
): Observable<NodeWithId<Text>> {
  return defer(() => {
    const node = document.createTextNode('');

    const decorators: Array<Decorator<NodeWithId<Text>>> = [applyTextContent(spec.textContent)];
    if (spec.decorators) {
      decorators.push(...spec.decorators);
    }

    return renderNode({
      ...spec,
      type: RenderSpecType.NODE,
      node,
      decorators,
    });
  });
}
