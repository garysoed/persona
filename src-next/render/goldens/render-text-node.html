<test-host><!-- #ref -->textContent</test-host>